### DISCUSSION TITLE
UIKit navigation

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#456] [General]
I'd like to create `ComposableNavigationController`, this discussion is meant to be a place for implementation ideas.

**Comment by maximkrouk** [comment]
The first step was to create recursive navigation.
> - I use [`DeclarativeConfiguration`](https://github.com/MakeupStudio/swift-declarative-configuration) framework for handlers
> - My controllers are subclasses of `ComposableViewController` which contains `ComposableCore { Store?, ViewStore? }` instance
> - CoreResettable protocol allows to set `controller.core` to `nil` by calling `controller.resetCore()`
> - `popToRootViewController` is not handled in the example
> - Originally this controller also should have some kind of a producer to let parent controller manage child states on navigation events
> - Probably this controller will be out of the composable application state, at least at first versions

```swift
class RecursiveNavigationController<
    RootController: UIViewController,
    RecursiveController: UIViewController
>: UINavigationController {
    var rootController: RootController! { viewControllers.first.as(RootController.self) }
    private(set) var recursiveControllers: [RecursiveController] = []
    
    @Handler<Void>
    var onRecursiveControllerPop
    
    convenience init(rootController: RootController) {
        self.init(rootViewController: rootController)
    }
    
    override init(rootViewController: UIViewController) {
        precondition(rootViewController is RootController, "Unexpected root viewController type")
        super.init(rootViewController: rootViewController)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    open override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        if
            let controller = viewController as? RecursiveController,
            !(viewControllers.isEmpty && RecursiveController.self == RootController.self)
        {
            recursiveControllers.append(controller)
        }
        if !viewControllers.isEmpty { viewController.setupBackButton() }
        super.pushViewController(viewController, animated: animated)
    }
    
    open override func popViewController(animated: Bool) -> UIViewController? {
        let controller = super.popViewController(animated: animated)
        if controller == recursiveControllers.last {
            _ = recursiveControllers.popLast()
            if let coordinator = transitionCoordinator, animated {
                coordinator.animate(alongsideTransition: nil) { _ in
                    self._onRecursiveControllerPop()
                    (controller as? CoreRessetable)?.resetCore()
                }
            } else {
                self._onRecursiveControllerPop()
                (controller as? CoreRessetable)?.resetCore()
            }
        }
        return controller
    }
    
    override func setViewControllers(_ viewControllers: [UIViewController], animated: Bool) {
        self.recursiveControllers = viewControllers.compactMap { $0 as? RecursiveController }
        super.setViewControllers(viewControllers, animated: animated)
    }
    
    func updateRecursiveControllersCount(to count: Int) {
        if recursiveControllers.count < count {
            while recursiveControllers.count + 1 < count {
                pushViewController(RecursiveController(), animated: false)
            }
            pushViewController(RecursiveController(), animated: true)
        } else if recursiveControllers.count > count {
            while recursiveControllers.count - 1 > count {
                _ = popViewController(animated: false)
            }
            _ = popViewController(animated: true)
        }
    }
}
```

**Reply by maximkrouk** [reply]
Parent controller should have an array of [ChildState] which is mutated after pop actions and additions/removals to/from that array should automatically push/pop ComposableChildControllers

**Reply by maximkrouk** [reply]
Non-composable controllers should probably be allowed too

**Comment by maximkrouk** [comment]
Second step is to understand how to represent `NavigationControllerState`

**Reply by maximkrouk** [reply]
`ComposableViewController<State, Action>` is typed, so maybe some kind of a ContainerViewController is needed to be created and then `containerVC.setStore<State, Action>()` will create `ComposableViewController<State, Action>`, set it's view to itself and add a new `ComposableViewController ` as child...

**Reply by maximkrouk** [reply]
Also for basic (non-recursive) navigation I use
```swift
public protocol Routable: Equatable {
    typealias RoutingAction = <#CurrentTarget#>.RoutingAction<Route>
    associatedtype Route: Equatable
    var currentRoute: Route? { get set }
}

public enum RoutingAction<Route: Equatable>: Equatable {
    case route(to: Route?)
    public static var dismiss: RoutingAction { .route(to: .none) }
}

extension Reducer where State: Routable, Action == State.RoutingAction, Environment == Void {
    public static func router() -> Reducer {
        .init { state, action, environment in
            switch action {
            case let .route(to: newRoute):
                state.currentRoute = newRoute
                return .none
            }
        }
    }
}

extension Reducer  {
    public static func router<LocalState: Routable>(
        for stateType: LocalState.Type
    ) -> Reducer<LocalState, LocalState.RoutingAction, Void> { .router() }
}
```

**Reply by maximkrouk** [reply]
Maybe it's a better way to manage navigation, but I'm not sure it fits recursive navigation, because states are not recursive

**Reply by maximkrouk** [reply]
Why do I need recursive navigation?
`Feed` → `Profile` → `Feed` → `Profile` → `Feed` → ...

I resolved it with
```
Main {
    feed: Feed
    recursiveProfiles: [Profile]
}

Profile {
    Feed
}
```

`Feed` has `showProfile` action which is handled by `Main` reducer

I could do indirect struct for feed and include profiles in it, but SPM will not allow to add `Feed` module into `Profile`, and `Profile` into `Feed` in the same time

**Comment by darrarski** [comment]
@maximkrouk this might be interesting for you:

I've developed a `UINavigationController` driven by `Store` with a help of new (not yet released) navigation API.

https://github.com/darrarski/store-navigation-controller

You can check out my post about it here:

https://github.com/pointfreeco/swift-composable-architecture/discussions/1674

**Reply by maximkrouk** [reply]
Hi, I'll take a closer look soon, on high level it looks interesting

I came up to a convenient solution for Combine-driven navigation, but it breaks modularization if you need recursive navigation

https://github.com/capturecontext/combine-cocoa-navigation

Sadly I haven't been using TCA for a while, had to move to another country and got some other difficulties in life, but just yesterday I got my sleep schedule back and felt some will to get back to some pet projects and open source culture, so hopefully there will be some updates to combine-navigation and TCA-extensions soon 😎

**Reply by maximkrouk** [reply]
I'm back to the TCA world, there were a bunch of updates and now I'm in process of updating my packages, here is a separate discussion for [CombineNavigation package](https://github.com/pointfreeco/swift-composable-architecture/discussions/2662), so I'm closing this one
